src/finalizer.rs

```diff
@@ -17,25 +18,45 @@ where
     };
 }
 
-/// Adds our finalizer to the list of finalizers.
-pub async fn add_finalizer<T>(client: Client, resource: &T, finalizer: &str) -> OperatorResult<T>
+/// This will add the passed finalizer to the list of finalizers for the resource if it doesn't exist yet
+/// and will update the resource in Kubernetes.
+///
+/// It'll return `true` if we changed the object in Kubernetes and `false` if no modification was needed.
+/// If the object is currently being deleted this _will_ return an Error!
+pub async fn add_finalizer<T>(
+    client: &Client,
+    resource: &T,
+    finalizer: &str,
+) -> OperatorResult<bool>
 where
-    T: Resource + Clone + Meta + DeserializeOwned,
+    T: Clone + Meta + DeserializeOwned,
 {
+    if has_finalizer(resource, finalizer) {
+        debug!("Finalizer [{}] already exists, continuing...", finalizer);
+
+        return Ok(false);
+    }
+
     let new_metadata = json!({
         "metadata": {
             "finalizers": [finalizer.to_string()]
         }
     });
-    client.merge_patch(resource, new_metadata).await
+    client.merge_patch(resource, new_metadata).await?;
+    Ok(true)
 }
 
 /// Removes our finalizer from a resource object.
 ///
 /// # Arguments
-/// `name` - is the name of the resource we want to patch
-/// `namespace` is the namespace of where the resource to patch lives
-pub async fn remove_finalizer<T>(client: Client, resource: &T, finalizer: &str) -> OperatorResult<T>
+///
```

Comment:
Missing client argument comment. I know its self explanatory but you commented it above. Just to be consistent.
/// - `client` - The Client to access Kubernetes